lib/debug/variable_inspector.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module DEBUGGER__
+  class VariableInspector
+    class Member
+      attr_reader :name, :value
+
+      def initialize(name:, value:, internal: false)
+        @name = name
+        @value = value
+        @is_internal = internal
+      end
+
+      def internal?
+        @is_internal
+      end
+
+      def self.internal name:, value:
+        new(name:, value:, internal: true)
+      end
+    end
+
+    def indexed_members_of obj, start:, count:
+      (start ... (start + count)).map { |i| Member.new(name: i.to_s, value: obj[i]) }
+    end
+
+    def named_members_of obj
+      members = case obj
+      when Array then obj.map.with_index { |o, i| Member.new(name: i.to_s, value: o) }
+      when Hash then obj.map { |k, v| Member.new(name: value_inspect(k), value: v) }
+      when Struct then obj.members.map { |name| Member.new(name:, value: obj[name]) }
+      when String
+        members = [
+          Member.internal(name: '#length', value: obj.length),
+          Member.internal(name: '#encoding', value: obj.encoding),
+        ]
+        if defined?(DEBUGGER__::NaiveString)
+          printed_str = value_inspect(obj)
+          members << Member.internal(name: "#dump", value: DEBUGGER__::NaiveString.new(obj)) if printed_str.end_with?('...')
+        end
+        members
+      when Class, Module then [Member.internal(name: "%ancestors", value: obj.ancestors[1..])]
+      when Range then [
+        Member.internal(name: "#begin", value: obj.begin),
+        Member.internal(name: "#end", value: obj.end),
+      ]
+      else []
+      end
+
+      unless defined?(DEBUGGER__::NaiveString) && DEBUGGER__::NaiveString === obj
+        members += M_INSTANCE_VARIABLES.bind_call(obj).sort.map{|iv|
+          Member.new(name: iv, value: M_INSTANCE_VARIABLE_GET.bind_call(obj, iv))
+        }
+        members.unshift  Member.internal(name: '#class', value: M_CLASS.bind_call(obj))
+      end
+
+      members
+    end
+
+    private
+
+    MAX_LENGTH = 180
+
+    def value_inspect obj, short: true
+      # TODO: max length should be configurable?
+      str = DEBUGGER__.safe_inspect obj, short: short, max_length: MAX_LENGTH
+
+      if str.encoding == Encoding::UTF_8
+        str.scrub
+      else
+        str.encode(Encoding::UTF_8, invalid: :replace, undef: :replace)
+      end
+    end
```

Comment:
I just copied this snippet for now, but my eventual goal is to move all presentation logic out of `ThreadClient` and into a shared class like this.